File: polyglot/php_worker/task_codec_rejection_probe.php
<?php

declare(strict_types=1);

use Composer\InstalledVersions;
use DurableWorkflow\Client;
use DurableWorkflow\Codec\AvroPayloadCodec;
use DurableWorkflow\Transport\Transport;
use DurableWorkflow\Worker;
use DurableWorkflow\Worker\ActivityContext;
use DurableWorkflow\Worker\QueryContext;
use DurableWorkflow\Worker\WorkflowContext;

$sourceAutoload = getenv('DURABLE_WORKFLOW_PHP_SDK_SOURCE_AUTOLOAD');
$autoload = is_string($sourceAutoload) && $sourceAutoload !== ''
    ? $sourceAutoload
    : (file_exists(__DIR__.'/vendor/autoload.php')
        ? __DIR__.'/vendor/autoload.php'
        : dirname(__DIR__, 2).'/vendor/autoload.php');
require $autoload;
require_once __DIR__.'/worker.php';

/** @return list<array{name: string, present: bool, value?: mixed}> */
function unsupportedCodecCases(): array
{
    return [
        ['name' => 'missing', 'present' => false],
        ['name' => 'empty', 'present' => true, 'value' => ''],
        ['name' => 'json', 'present' => true, 'value' => 'json'],
        ['name' => 'unknown', 'present' => true, 'value' => 'custom'],
        ['name' => 'wrong_case', 'present' => true, 'value' => 'Avro'],
        ['name' => 'null', 'present' => true, 'value' => null],
        ['name' => 'non_string', 'present' => true, 'value' => ['avro']],
        ['name' => 'malformed', 'present' => true, 'value' => "avro\0"],
    ];
}

/**
 * @param  array{name: string, present: bool, value?: mixed}  $codecCase
 * @param  array<string, mixed>  $arguments
 * @return array<string, mixed>
 */
function sdkProbeTask(string $path, array $codecCase, array $arguments): array
{
    $task = match ($path) {
        'workflow' => [
            'task_id' => 'probe-workflow',
            'workflow_task_attempt' => 1,
            'lease_owner' => 'probe-worker',
            'workflow_id' => 'probe-workflow-id',
            'run_id' => 'probe-workflow-run',
            'workflow_type' => 'probe.workflow',
            'arguments' => $arguments,
            'history_events' => [],
        ],
        'update' => [
            'task_id' => 'probe-update',
            'workflow_task_attempt' => 1,
            'lease_owner' => 'probe-worker',
            'workflow_id' => 'probe-update-id',
            'run_id' => 'probe-update-run',
            'workflow_type' => 'probe.workflow',
            'workflow_update_id' => 'probe-update-command',
            'history_events' => [[
                'event_type' => 'UpdateAccepted',
                'payload' => [
                    'update_id' => 'probe-update-command',
                    'update_name' => 'increment',
                    'arguments' => $arguments,
                ],
            ]],
        ],
        'activity' => [
            'task_id' => 'probe-activity',
            'activity_attempt_id' => 'probe-activity-attempt',
            'lease_owner' => 'probe-worker',
            'activity_type' => 'probe.activity',
            'arguments' => $arguments,
        ],
        'query' => [
            'query_task_id' => 'probe-query',
            'query_task_attempt' => 1,
            'lease_owner' => 'probe-worker',
            'workflow_id' => 'probe-query-id',
            'run_id' => 'probe-query-run',
            'workflow_type' => 'probe.workflow',
            'query_name' => 'status',
            'query_arguments' => $arguments,
            'history_events' => [],
        ],
        default => throw new InvalidArgumentException("Unknown SDK probe path {$path}."),
    };
    if ($codecCase['present']) {
        $task['payload_codec'] = $codecCase['value'] ?? null;
    }

    return $task;
}

/** @return object&Transport */
function probeTransport(string $path, array $task): Transport
{
    return new class($path, $task) implements Transport
    {
        /** @var list<array{method: string, uri: string, body: ?array<string, mixed>}> */
        public array $requests = [];

        private bool $delivered = false;

        /** @param array<string, mixed> $task */
        public function __construct(private readonly string $path, private readonly array $task) {}

        public function send(string $method, string $uri, array $headers, ?array $body = null): ?array
        {
            $this->requests[] = compact('method', 'uri', 'body');
            if (str_ends_with($uri, '/workflow-tasks/poll')) {
                if (! $this->delivered && in_array($this->path, ['workflow', 'update'], true)) {
                    $this->delivered = true;

                    return ['poll_status' => 'leased', 'task' => $this->task];
                }

                return ['poll_status' => 'empty', 'task' => null];
            }
            if (str_ends_with($uri, '/activity-tasks/poll')) {
                if (! $this->delivered && $this->path === 'activity') {
                    $this->delivered = true;

                    return ['poll_status' => 'leased', 'task' => $this->task];
                }

                return ['poll_status' => 'empty', 'task' => null];
            }
            if (str_ends_with($uri, '/query-tasks/poll')) {
                if (! $this->delivered && $this->path === 'query') {
                    $this->delivered = true;

                    return ['poll_status' => 'leased', 'task' => $this->task];
                }

                return ['poll_status' => 'empty', 'task' => null];
            }
            if (str_ends_with($uri, '/heartbeat')) {
                return [
                    'task_id' => $this->task['task_id'],
                    'workflow_task_attempt' => $this->task['workflow_task_attempt'],
                    'lease_owner' => $this->task['lease_owner'],
                    'renewed' => true,
                    'reason' => null,
                ];
            }
            if (str_ends_with($uri, '/complete') || str_ends_with($uri, '/fail')) {
                return ['acknowledged' => true];
            }

            throw new RuntimeException("Unexpected probe request: {$method} {$uri}");
        }
    };
}

/**
 * @param  array<string, mixed>  $task
 * @param  object&Transport  $transport
 */
function sdkProbeWorker(array $task, Transport $transport, int &$handlerCalls): Worker
{
    $worker = new Worker(
        new Client('http://probe.invalid', transport: $transport),
        'probe-queue',
        workerId: 'probe-worker',
    );
    $worker
        ->registerWorkflow(
            'probe.workflow',
            static function (WorkflowContext $context, mixed $input = null) use (&$handlerCalls): string {
                $handlerCalls++;

                return 'workflow-complete';
            },
        )
        ->registerUpdate(
            'probe.workflow',
            'increment',
            static function (QueryContext $context, int $value) use (&$handlerCalls): int {
                $handlerCalls++;

                return $value + 1;
            },
        )
        ->registerActivity(
            'probe.activity',
            static function (ActivityContext $context, mixed $input = null) use (&$handlerCalls): string {
                $handlerCalls++;

                return 'activity-complete';
            },
        )
        ->registerQuery(
            'probe.workflow',
            'status',
            static function (QueryContext $context, mixed $input = null) use (&$handlerCalls): string {
                $handlerCalls++;

                return 'query-complete';
            },
        );

    return $worker;
}

/** @param list<array{method: string, uri: string, body: ?array<string, mixed>}> $requests */
function requestEvidence(array $requests): array
{
    $failures = array_values(array_filter(
        $requests,
        static fn (array $request): bool => str_ends_with($request['uri'], '/fail'),
    ));
    $completions = array_values(array_filter(
        $requests,
        static fn (array $request): bool => str_ends_with($request['uri'], '/complete'),
    ));
    $failureDocument = json_encode($failures[0]['body'] ?? null, JSON_THROW_ON_ERROR);

    return [
        'failure_count' => count($failures),
        'completion_count' => count($completions),
        'unsupported_diagnostic' => str_contains($failureDocument, 'unsupported_payload_codec'),
        'decode_diagnostic' => str_contains($failureDocument, 'invalid_payload_framing'),
    ];
}

/** @return array<string, mixed> */
function runSdkRejectionProbe(string $path, array $codecCase): array
{
    $task = sdkProbeTask(
        $path,
        $codecCase,
        ['codec' => 'avro', 'blob' => 'decode-must-not-run'],
    );
    $transport = probeTransport($path, $task);
    $handlerCalls = 0;
    $handled = sdkProbeWorker($task, $transport, $handlerCalls)->tick(0);
    $requests = requestEvidence($transport->requests);
    $passed = $handled
        && $handlerCalls === 0
        && $requests['failure_count'] === 1
        && $requests['completion_count'] === 0
        && $requests['unsupported_diagnostic']
        && ! $requests['decode_diagnostic'];

    return [
        'runtime' => 'php',
        'worker' => 'sdk',
        'path' => $path,
        'codec_case' => $codecCase['name'],
        'status' => $passed ? 'rejected_before_decode_or_handler' : 'failed',
        'handler_calls' => $handlerCalls,
        ...$requests,
    ];
}

/** @return array<string, mixed> */
function runSdkValidProbe(string $path): array
{
    $arguments = (new AvroPayloadCodec)->envelope([$path === 'update' ? 41 : 'input']);
    $task = sdkProbeTask($path, ['name' => 'avro', 'present' => true, 'value' => 'avro'], $arguments);
    $transport = probeTransport($path, $task);
    $handlerCalls = 0;
    $handled = sdkProbeWorker($task, $transport, $handlerCalls)->tick(0);
    $requests = requestEvidence($transport->requests);
    $passed = $handled
        && $handlerCalls === 1
        && $requests['failure_count'] === 0
        && $requests['completion_count'] === 1;

    return [
        'runtime' => 'php',
        'worker' => 'sdk',
        'path' => $path,
        'codec_case' => 'avro',
        'status' => $passed ? 'decoded_and_handled' : 'failed',
        'handler_calls' => $handlerCalls,
        ...$requests,
    ];
}

/** @return object&Transport */
function manualProbeTransport(): Transport
{
    return new class implements Transport
    {
        /** @var list<array{method: string, uri: string, body: ?array<string, mixed>}> */
        public array $requests = [];

        public function send(string $method, string $uri, array $headers, ?array $body = null): ?array
        {
            $this->requests[] = compact('method', 'uri', 'body');

            return ['acknowledged' => true];
        }
    };
}

/** @return array<string, mixed> */
function runManualProbe(string $path, array $codecCase, bool $valid): array
{
    $transport = manualProbeTransport();
    $client = new Client('http://probe.invalid', transport: $transport);
    $arguments = $valid
        ? $client->payloadCodec()->envelope([['codec' => 'customer-owned', 'payload_codec' => 'metadata']])
        : ['codec' => 'avro', 'blob' => 'decode-must-not-run'];
    $task = [
        'task_id' => 'manual-probe',
        'query_task_id' => 'manual-probe',
        'activity_attempt_id' => 'manual-probe-attempt',
        'lease_owner' => 'manual-probe-worker',
        'activity_type' => 'polyglot.php.marker',
        'workflow_id' => 'manual-probe-workflow',
        'run_id' => 'manual-probe-run',
        'arguments' => $arguments,
        'workflow_arguments' => $arguments,
        'history_events' => [],
    ];
    if ($codecCase['present']) {
        $task['payload_codec'] = $codecCase['value'] ?? null;
    }
    if ($path === 'activity') {
        handleManualActivityTask($client, 'manual-probe-worker', $task);
    } else {
        handleManualQueryTask($client, 'manual-probe-worker', $task, $client->payloadCodec());
    }
    $requests = requestEvidence($transport->requests);
    $passed = $valid
        ? $requests['completion_count'] === 1 && $requests['failure_count'] === 0
        : $requests['completion_count'] === 0
            && $requests['failure_count'] === 1
            && $requests['unsupported_diagnostic']
            && ! $requests['decode_diagnostic'];

    return [
        'runtime' => 'php',
        'worker' => 'manual',
        'path' => $path,
        'codec_case' => $codecCase['name'],
        'status' => $passed
            ? ($valid ? 'decoded_and_handled' : 'rejected_before_decode_or_handler')
            : 'failed',
        ...$requests,
    ];
}

$rejections = [];
$validControls = [];
foreach (unsupportedCodecCases() as $codecCase) {
    foreach (['workflow', 'update', 'activity', 'query'] as $path) {
        $rejections[] = runSdkRejectionProbe($path, $codecCase);
    }
    foreach (['activity', 'query'] as $path) {
        $rejections[] = runManualProbe($path, $codecCase, false);
    }
}
foreach (['workflow', 'update', 'activity', 'query'] as $path) {
    $validControls[] = runSdkValidProbe($path);
}
foreach (['activity', 'query'] as $path) {
    $validControls[] = runManualProbe(
        $path,
        ['name' => 'avro', 'present' => true, 'value' => 'avro'],
        true,
    );
}
$failed = array_values(array_filter(
    [...$rejections, ...$validControls],
    static fn (array $outcome): bool => $outcome['status'] === 'failed',
));
$evidence = [
    'schema' => 'durable-workflow.sample-app.task-codec-rejection-probe',
    'version' => 1,
    'runtime' => 'php',
    'artifact' => [
        'name' => 'durable-workflow/sdk',
        'version' => InstalledVersions::getPrettyVersion('durable-workflow/sdk')
            ?? InstalledVersions::getVersion('durable-workflow/sdk'),
    ],
    'rejection_outcomes' => $rejections,
    'valid_controls' => $validControls,
    'summary' => [
        'status' => $failed === [] ? 'passed' : 'failed',
        'rejection_count' => count($rejections),
        'valid_control_count' => count($validControls),
        'failed_count' => count($failed),
    ],
];
fwrite(STDOUT, json_encode($evidence, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR)."\n");
exit($failed === [] ? 0 : 1);

File: polyglot/python_worker/scripts/task_codec_rejection_probe.py
"""Exercise published Python worker task boundaries with hostile root codec tags."""
from __future__ import annotations

import asyncio
import importlib.metadata
import json
import sys
from pathlib import Path
from typing import Any

from durable_workflow import Worker, activity, serializer, workflow

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
import activities as manual_activities


@workflow.defn(name="probe.workflow")
class ProbeWorkflow:
    workflow_calls = 0
    query_calls = 0

    def run(self, _context: object, value: object = None) -> object:
        type(self).workflow_calls += 1
        return value

    @workflow.query("status")
    def status(self, value: object = None) -> object:
        type(self).query_calls += 1
        return value


ACTIVITY_CALLS = 0


@activity.defn(name="probe.activity")
def probe_activity(value: object = None) -> object:
    global ACTIVITY_CALLS
    ACTIVITY_CALLS += 1
    return value


class ProbeClient:
    external_storage = None
    external_storage_threshold_bytes = None
    external_storage_cache = None
    metrics = None
    payload_size_warning_config = None
    namespace = "default"

    def __init__(self) -> None:
        self.workflow_failures: list[dict[str, Any]] = []
        self.workflow_completions: list[dict[str, Any]] = []
        self.activity_failures: list[dict[str, Any]] = []
        self.activity_completions: list[dict[str, Any]] = []
        self.query_failures: list[dict[str, Any]] = []
        self.query_completions: list[dict[str, Any]] = []

    async def fail_workflow_task(self, **arguments: Any) -> dict[str, bool]:
        self.workflow_failures.append(arguments)
        return {"acknowledged": True}

    async def complete_workflow_task(self, **arguments: Any) -> dict[str, bool]:
        self.workflow_completions.append(arguments)
        return {"acknowledged": True}

    async def fail_activity_task(self, **arguments: Any) -> dict[str, bool]:
        self.activity_failures.append(arguments)
        return {"acknowledged": True}

    async def complete_activity_task(self, **arguments: Any) -> dict[str, bool]:
        self.activity_completions.append(arguments)
        return {"acknowledged": True}

    async def fail_query_task(self, **arguments: Any) -> dict[str, bool]:
        self.query_failures.append(arguments)
        return {"acknowledged": True}

    async def complete_query_task(self, **arguments: Any) -> dict[str, bool]:
        self.query_completions.append(arguments)
        return {"acknowledged": True}


CODEC_CASES: tuple[tuple[str, bool, object], ...] = (
    ("missing", False, None),
    ("empty", True, ""),
    ("json", True, "json"),
    ("unknown", True, "custom"),
    ("wrong_case", True, "Avro"),
    ("null", True, None),
    ("non_string", True, ["avro"]),
    ("malformed", True, "avro\0"),
)


def task(path: str, present: bool, codec: object, arguments: object) -> dict[str, Any]:
    if path == "workflow":
        value = {
            "task_id": "probe-workflow",
            "workflow_task_attempt": 1,
            "lease_owner": "probe-worker",
            "workflow_id": "probe-workflow-id",
            "run_id": "probe-workflow-run",
            "workflow_type": "probe.workflow",
            "arguments": arguments,
            "history_events": [],
        }
    elif path == "activity":
        value = {
            "task_id": "probe-activity",
            "activity_attempt_id": "probe-activity-attempt",
            "lease_owner": "probe-worker",
            "activity_type": "probe.activity",
            "arguments": arguments,
        }
    elif path == "query":
        value = {
            "query_task_id": "probe-query",
            "query_task_attempt": 1,
            "lease_owner": "probe-worker",
            "workflow_id": "probe-query-id",
            "run_id": "probe-query-run",
            "workflow_type": "probe.workflow",
            "query_name": "status",
            "workflow_arguments": arguments,
            "query_arguments": arguments,
            "history_events": [],
        }
    else:
        raise ValueError(f"unknown Python SDK probe path {path!r}")
    if present:
        value["payload_codec"] = codec
    return value


def worker(client: ProbeClient) -> Worker:
    return Worker(
        client,  # type: ignore[arg-type]
        task_queue="probe-queue",
        worker_id="probe-worker",
        workflows=[ProbeWorkflow],
        activities=[probe_activity],
    )


def diagnostic(failures: list[dict[str, Any]]) -> tuple[bool, bool]:
    document = json.dumps(failures, default=repr)
    return "unsupported_payload_codec" in document, "decode-must-not-run" in document


async def run_sdk_path(path: str, payload: dict[str, Any]) -> None:
    runtime = worker(payload["client"])
    if path == "workflow":
        await runtime._run_workflow_task(payload["task"])
    elif path == "activity":
        await runtime._run_activity_task(payload["task"])
    else:
        await runtime._run_query_task(payload["task"])


def sdk_observation(path: str, client: ProbeClient) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    if path == "workflow":
        return client.workflow_failures, client.workflow_completions
    if path == "activity":
        return client.activity_failures, client.activity_completions
    return client.query_failures, client.query_completions


async def sdk_rejection(path: str, name: str, present: bool, codec: object) -> dict[str, Any]:
    global ACTIVITY_CALLS
    ACTIVITY_CALLS = 0
    ProbeWorkflow.workflow_calls = 0
    ProbeWorkflow.query_calls = 0
    client = ProbeClient()
    hostile = task(
        path,
        present,
        codec,
        {"codec": "avro", "blob": "decode-must-not-run"},
    )
    await run_sdk_path(path, {"client": client, "task": hostile})
    failures, completions = sdk_observation(path, client)
    unsupported, poison_observed = diagnostic(failures)
    handler_calls = ACTIVITY_CALLS + ProbeWorkflow.workflow_calls + ProbeWorkflow.query_calls
    passed = (
        len(failures) == 1
        and not completions
        and unsupported
        and not poison_observed
        and handler_calls == 0
    )
    return {
        "runtime": "python",
        "worker": "sdk",
        "path": path,
        "codec_case": name,
        "status": "rejected_before_decode_or_handler" if passed else "failed",
        "failure_count": len(failures),
        "completion_count": len(completions),
        "unsupported_diagnostic": unsupported,
        "poison_decode_diagnostic": poison_observed,
        "handler_calls": handler_calls,
    }


async def sdk_valid(path: str) -> dict[str, Any]:
    global ACTIVITY_CALLS
    ACTIVITY_CALLS = 0
    ProbeWorkflow.workflow_calls = 0
    ProbeWorkflow.query_calls = 0
    client = ProbeClient()
    arguments = serializer.envelope(["input"], codec="avro")
    await run_sdk_path(path, {"client": client, "task": task(path, True, "avro", arguments)})
    failures, completions = sdk_observation(path, client)
    handler_calls = ACTIVITY_CALLS + ProbeWorkflow.workflow_calls + ProbeWorkflow.query_calls
    passed = not failures and len(completions) == 1 and handler_calls > 0
    return {
        "runtime": "python",
        "worker": "sdk",
        "path": path,
        "codec_case": "avro",
        "status": "decoded_and_handled" if passed else "failed",
        "failure_count": len(failures),
        "completion_count": len(completions),
        "handler_calls": handler_calls,
    }


async def manual_rejection(name: str, present: bool, codec: object) -> dict[str, Any]:
    client = ProbeClient()
    manual_task = {
        "task_id": "manual-probe",
        "activity_attempt_id": "manual-probe-attempt",
        "activity_type": manual_activities.PYTHON_TYPED_ERROR_ACTIVITY,
        "arguments": {"codec": "avro", "blob": "decode-must-not-run"},
    }
    if present:
        manual_task["payload_codec"] = codec
    outcome = await manual_activities.handle_typed_error_task(
        client,  # type: ignore[arg-type]
        "manual-probe-worker",
        manual_task,
    )
    unsupported, poison_observed = diagnostic(client.activity_failures)
    passed = (
        outcome == "unsupported_payload_codec"
        and len(client.activity_failures) == 1
        and not client.activity_completions
        and unsupported
        and not poison_observed
        and "details" not in client.activity_failures[0]
    )
    return {
        "runtime": "python",
        "worker": "manual",
        "path": "activity",
        "codec_case": name,
        "status": "rejected_before_decode_or_handler" if passed else "failed",
        "failure_count": len(client.activity_failures),
        "completion_count": len(client.activity_completions),
        "unsupported_diagnostic": unsupported,
        "poison_decode_diagnostic": poison_observed,
    }


async def manual_valid() -> dict[str, Any]:
    client = ProbeClient()
    customer_payload = {"codec": "json", "payload_codec": "customer-owned"}
    outcome = await manual_activities.handle_typed_error_task(
        client,  # type: ignore[arg-type]
        "manual-probe-worker",
        {
            "task_id": "manual-probe",
            "activity_attempt_id": "manual-probe-attempt",
            "activity_type": manual_activities.PYTHON_TYPED_ERROR_ACTIVITY,
            "payload_codec": "avro",
            "arguments": serializer.envelope([customer_payload], codec="avro"),
        },
    )
    handled_failure = client.activity_failures[0] if client.activity_failures else {}
    request = (((handled_failure.get("details") or {}).get("structured") or {}).get("request"))
    passed = outcome == "handled" and request == customer_payload
    return {
        "runtime": "python",
        "worker": "manual",
        "path": "activity",
        "codec_case": "avro",
        "status": "decoded_and_handled" if passed else "failed",
        "handler_calls": 1 if passed else 0,
    }


async def main() -> int:
    rejections = []
    for name, present, codec in CODEC_CASES:
        for path in ("workflow", "activity", "query"):
            rejections.append(await sdk_rejection(path, name, present, codec))
        rejections.append(await manual_rejection(name, present, codec))
    valid_controls = [await sdk_valid(path) for path in ("workflow", "activity", "query")]
    valid_controls.append(await manual_valid())
    failures = [item for item in [*rejections, *valid_controls] if item["status"] == "failed"]
    evidence = {
        "schema": "durable-workflow.sample-app.task-codec-rejection-probe",
        "version": 1,
        "runtime": "python",
        "artifact": {
            "name": "durable-workflow",
            "version": importlib.metadata.version("durable-workflow"),
        },
        "rejection_outcomes": rejections,
        "valid_controls": valid_controls,
        "summary": {
            "status": "passed" if not failures else "failed",
            "rejection_count": len(rejections),
            "valid_control_count": len(valid_controls),
            "failed_count": len(failures),
        },
    }
    print(json.dumps(evidence, indent=2, sort_keys=True))
    return 0 if not failures else 1


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
